src/components/modebar/index.js

```diff
@@ -97,18 +121,35 @@ proto.createButton = function (config) {
     button.setAttribute('rel', 'tooltip');
     button.className = 'modebar-btn';
 
-    if (config.attr !== undefined) button.setAttribute('data-attr', config.attr);
-    if (config.val !== undefined) button.setAttribute('data-val', config.val);
-    button.setAttribute('data-title', config.title);
-    button.setAttribute('data-gravity', config.gravity || 'n');
-    button.addEventListener('click', function () {
-            config.click.apply(_this, arguments);
+    var title = config.title;
+    if(title !== null && title !== false && title !== '') {
+        button.setAttribute('data-title', title || config.name);
+    }
+
+    if(config.attr !== undefined) button.setAttribute('data-attr', config.attr);
+
+    var val = config.val;
+    if(val !== undefined) {
+        if(typeof val === 'function') val = val(this.graphInfo);
+        button.setAttribute('data-val', val);
+    }
+
+    var click = config.click;
+    if(typeof click !== 'function') {
+        throw new Error('must provide button \'click\' function in button config');
+    }
+    else {
+        button.addEventListener('click', function(ev) {
+            config.click(_this.graphInfo, ev);
+            _this.updateActiveButton(ev.currentTarget);
```

Comment:
**IMPORTANT**

Now the click handler are passed `gd` and `ev` (the event object) only. In other words, the click handlers know nothing about `ModeBar`.

`updateActiveButton` is called via `Plotly.relayout` (which calls the mode bar manager which calls `modeBar.updateActiveButton` on update). So this `updateActiveButton` is redundant for the `dragmode` and `hovermode` handlers.

But, the geo hover toggle button needs the above line,  because it can't call relayout at the moment (relayout would reset the camera everytime the geo `hovermode` is changed).

We shouldn't be worried about looping through <10 buttons twice. Correct?